Sources/TensorFlow/Operators/Basic.swift

```diff
@@ -375,6 +426,103 @@ internal extension Tensor where Scalar: TensorFlowFloatingPoint {
+        let posAxis = axis < 0 ? axis + rank : axis
+        
+        // We have a fast gradient implementation for the case when `posAxis == 0`.
+        if posAxis == 0 {
+            return (result, { [shape = shapeTensor] v in
+                let indicesCount = indices.scalarCountTensor.rankLifted()
+                let valuesShape = Tensor<Int32>(concatenating: [indicesCount, shape[1...]])
+                let values = v.reshaped(toShape: valuesShape)
+                let valueIndices = indices.reshaped(toShape: indicesCount)
+                return Raw.unsortedSegmentSum(
+                    data: values,
+                    segmentIds: valueIndices,
+                    numSegments: shape[0])
+            })
+        }
+        
+        return (result, { [shape = shapeTensor] v in
+            let indicesSize = Tensor<Int32>(Int32(indices.scalarCount)).rankLifted()
+            let outerShape = shape[..<posAxis]
+            let outerSize = outerShape.scalarCount
+            let innerShape = shape[(posAxis + 1)...]
+            let innerSize = innerShape.scalarCount
+            let outerIndices = Tensor<Int32>(rangeFrom: 0, to: Int32(outerSize), stride: 1)
+            let innerIndices = Tensor<Int32>(
+                rangeFrom: Int32(outerSize) + 1,
+                to: Int32(outerSize) + 1 + Int32(innerSize),
+                stride: 1)
+            let valuesShape = Tensor<Int32>(concatenating: [outerShape, indicesSize, innerShape])
+            let values = v.reshaped(toShape: valuesShape)
+            let valueIndices = indices.reshaped(toShape: indicesSize)
+
+            // We need to sum up every slice `values[..., i, ....]` corresponding to
+            // `tensor[..., indices[i], ...]`. Since `unsortedSegmentSum` does not support an axis
+            // parameter, we transpose the gather dimension to the front, then use
+            // `unsortedSegmentSum` to build a `[gatherAxis, outerAxes, innerAxes]` tensor with all
+            // the gradients affecting each index in `gatherAxis` summed up.
+            let permutations = Tensor<Int32>(concatenating: [
+                Tensor<Int32>([Int32(outerSize)]),
+                outerIndices,
+                innerIndices])
+            let transposedValues = values.transposed(withPermutations: permutations)
+            let gradient = Raw.unsortedSegmentSum(
+                data: transposedValues,
+                segmentIds: valueIndices,
+                numSegments: shape[posAxis])
+
+            // Finally, we invert the above transpose operation by moving dimension 0 back to its
+            // original position.
+            let inversePermutations = Tensor<Int32>(concatenating: [
+                outerIndices + 1,
+                Tensor<Int32>([0]),
+                innerIndices])
+            return gradient.transposed(withPermutations: inversePermutations)
+        })
+    }
+}
+
+public extension Tensor {
```

Comment:
I think that this API should be kept internal until we find a good spelling/semantics for it. The fact that it applies to `Tensor<Bool>` makes the `nonZero` name weird.

Reply:
We could split it in two functions: `nonZeroIndices` and `falseIndices` or something like that. How does that sound?

Reply:
Linguistically those adjectives are attached to "indices", which is not the intended meaning and could cause confusion. Also, which Python TF operator does it correspond to?

Reply:
It's part of the `tf.where` functionality when no `x` and `y` are provided (i.e., only `condition` is provided).

Reply:
It seems this is not being used by the implementation of `gathering`. Could we defer this to later PRs instead?

Reply:
That's true. I think it's used by one of the other PRs so I'll just move it there.